Framework/API/src/FunctionFactory.cpp

```diff
@@ -266,10 +266,28 @@ void FunctionFactoryImpl::inputError(const std::string &str) const {
 void FunctionFactoryImpl::addConstraints(IFunction_sptr fun,
                                          const Expression &expr) const {
   if (expr.name() == ",") {
-    for (const auto &constraint : expr) {
-      addConstraint(fun, constraint);
+    for (auto it = expr.begin(); it != expr.end(); ++it) {
+      // If this is a penalty term, we used it on the previous iteration
+      // so we move on to the next term.
+      auto constraint = (*it);
+      std::string constraint_term = constraint.terms()[0].str();
+      if (constraint_term.compare("penalty") == 0) {
+        continue;
+      }
+
+      if ((it + 1) != expr.end()) {
+        auto next_constraint = *(it + 1);
+        std::string next_term = next_constraint.terms()[0].str();
+        if (next_term.compare("penalty") == 0) {
+          addConstraint(fun, constraint, next_constraint);
+        } else {
+          addConstraint(fun, constraint);
+        }
+      } else {
+        addConstraint(fun, constraint);
+      }
```

Comment:
This needs a unit test.

Can this also be mentioned in the documentation somewhere? Release notes?

Reply:
Added unit tests and updated documentation.